File: src/lib/api/skinFixer.ts
import { invokeCommand } from './core';

/** One entry in the fix catalog (from Hematite's config). */
export interface FixEntry {
    id: string;
    name: string;
    description: string;
    severity: 'low' | 'medium' | 'high' | 'critical' | string;
    enabled: boolean;
    wad_level: boolean;
}

/** A single fix that fired (detected on scan, or applied on run). */
export interface FixOutcome {
    fix_id: string;
    fix_name: string;
    changes: number;
    file_path: string;
}

/** Per-project result of a scan or run. */
export interface ProjectFixReport {
    project: string;
    outcomes: FixOutcome[];
    fixes_applied: number;
    fixes_failed: number;
    files_removed: number;
    errors: string[];
    champion: string | null;
    skin_number: number | null;
    /** Set when the whole project failed (e.g. missing hashes). */
    error: string | null;
}

/** Progress event payload (`hematite-fix-progress`). */
export interface FixProgress {
    project: string;
    stage?: string;
    fix?: string;
    count?: number;
    note?: string;
}

/** The full fix catalog. */
export function hematiteListFixes(): Promise<FixEntry[]> {
    return invokeCommand<FixEntry[]>('hematite_list_fixes');
}

/** Detect-only scan of each project — reports which selected fixes fire, writes nothing. */
export function hematiteScanProjects(
    projectPaths: string[],
    fixIds: string[],
): Promise<ProjectFixReport[]> {
    return invokeCommand<ProjectFixReport[]>('hematite_scan_projects', { projectPaths, fixIds });
}

/** Apply the selected fixes; streams progress via the `hematite-fix-progress` event. */
export function hematiteRunFixes(
    projectPaths: string[],
    fixIds: string[],
    useLive: boolean,
): Promise<ProjectFixReport[]> {
    return invokeCommand<ProjectFixReport[]>('hematite_run_fixes', { projectPaths, fixIds, useLive });
}

File: src-tauri/src/commands/editor/skin_fixer.rs
//! Skin Fixer commands — drive Hematite's fix engine over Flint projects.
//!
//! `hematite_list_fixes` returns the fix catalog; `hematite_scan_projects` runs
//! a detect-only pass (which fixes fire per project, writes nothing);
//! `hematite_run_fixes` applies the selected fixes and streams progress via the
//! `hematite-fix-progress` event. All heavy work runs on a blocking thread — the
//! engine is sync + CPU/IO heavy.

use hematite_flint::{FixEntry, ProjectFixReport};
use tauri::{AppHandle, Emitter};

/// A `(project_dir, label)` pair. The label is what the report is keyed by in
/// the UI; default it to the dir when the frontend doesn't supply one.
fn pairs(project_paths: Vec<String>) -> Vec<(String, String)> {
    project_paths
        .into_iter()
        .map(|p| {
            let label = std::path::Path::new(&p)
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_else(|| p.clone());
            (p, label)
        })
        .collect()
}

#[tauri::command]
pub async fn hematite_list_fixes() -> Result<Vec<FixEntry>, String> {
    tokio::task::spawn_blocking(hematite_flint::list_available_fixes)
        .await
        .map_err(|e| format!("skin-fixer task failed: {e}"))
}

#[tauri::command]
pub async fn hematite_scan_projects(
    project_paths: Vec<String>,
    fix_ids: Vec<String>,
) -> Result<Vec<ProjectFixReport>, String> {
    let projects = pairs(project_paths);
    tokio::task::spawn_blocking(move || hematite_flint::scan_projects(&projects, &fix_ids))
        .await
        .map_err(|e| format!("skin-fixer scan failed: {e}"))
}

#[tauri::command]
pub async fn hematite_run_fixes(
    app: AppHandle,
    project_paths: Vec<String>,
    fix_ids: Vec<String>,
    use_live: bool,
) -> Result<Vec<ProjectFixReport>, String> {
    let projects = pairs(project_paths);
    tokio::task::spawn_blocking(move || {
        hematite_flint::run_projects(&projects, &fix_ids, use_live, |progress| {
            let _ = app.emit("hematite-fix-progress", progress);
        })
    })
    .await
    .map_err(|e| format!("skin-fixer run failed: {e}"))
}
